numpy/core/src/multiarray/methods.c

```diff
@@ -617,23 +617,31 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
 
     fd = npy_PyFile_Dup2(file, "wb", &orig_pos);
     if (fd == NULL) {
-        goto fail;
-    }
-    if (PyArray_ToFile(self, fd, sep, format) < 0) {
-        goto fail;
+        Py_DECREF(file);
+        return NULL;
     }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
-        goto fail;
+
+    int res_write = PyArray_ToFile(self, fd, sep, format);
+    PyObject *err_type, *err_value, *err_traceback;
+    PyErr_Fetch(&err_type, &err_value, &err_traceback);
+
+    int res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res_close < 0) {
+        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
     }
+
     if (own && npy_PyFile_CloseFile(file) < 0) {
```

Comment:
I think `npy_PyFile_CloseFile` is also unsafe to call here - you need to fetch the exception _again_, in case `npy_PyFile_DupClose2` created one. See my comment in https://github.com/numpy/numpy/pull/17598#issuecomment-713902244, this becomes much easier to follow if you extract a helper function.

Reply:
Thanks @eric-wieser, appreciate your patience. I didn't quite manage to refactor it into a helper function, but have updated it so that is now hopefully exception safe, and not too horrific to read.